plugins/com.google.cloud.tools.eclipse.dataflow.core/src/com/google/cloud/tools/eclipse/dataflow/core/launcher/PipelineLaunchConfiguration.java

```diff
@@ -122,20 +123,44 @@ public void setUseDefaultLaunchOptions(boolean useDefaultLaunchOptions) {
     this.useDefaultLaunchOptions = useDefaultLaunchOptions;
   }
 
-  public String getEclipseProjectName() {
-    return eclipseProjectName.orNull();
+  public MajorVersion getMajorVersion() {
+    return majorVersion;
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    if (this == obj) {
+      return true;
+    } else if (obj == null || getClass() != obj.getClass()) {
+      return false;
+    }
+    PipelineLaunchConfiguration other = (PipelineLaunchConfiguration) obj;
+    return Objects.equals(majorVersion, other.majorVersion)
+        && Objects.equals(argumentValues, other.argumentValues)
+        && Objects.equals(runner, other.runner)
+        && Objects.equals(userOptionsName, other.userOptionsName);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(majorVersion, argumentValues, runner, userOptionsName);
   }
 
-  private void setEclipseProjectName(ILaunchConfiguration configuration) throws CoreException {
-    eclipseProjectName = Optional.fromNullable(
-        configuration.getAttribute(
-            IJavaLaunchConfigurationConstants.ATTR_PROJECT_NAME, (String) null));
+  @Override
+  public String toString() {
+    return MoreObjects.toStringHelper(getClass())
+        .add("runnerName", runner != null ? runner.getRunnerName() : "(default runner)")
+        .add("argumentValues", argumentValues).toString();
   }
 
   private void setValuesFromLaunchConfiguration(ILaunchConfiguration configuration)
```

Comment:
Noticed this should just have been a constructor:
```java
  private PipelineLaunchConfiguration(MajorVersion majorVersion, ILaunchConfiguration configuration)
      throws CoreException {
    this.majorVersion = Preconditions.checkNotNull(majorVersion);

    // never set a default runner unless actually specified
```
I was hunting for a while who can call this when, but if we make this a constructor, it becomes obvious. `fromLaunchConfiguration` then could simply be
```java
  public static PipelineLaunchConfiguration fromLaunchConfiguration(MajorVersion majorVersion,
      ILaunchConfiguration launchConfiguration) throws CoreException {
    return new PipelineLaunchConfiguration(majorVersion, launchConfiguration);
  }
```
Maybe we should just ditch the static `fromLaunchConfiguration()`. This another-static-constructor method doesn't help me understand code, IMO.

But then, this refactoring doesn't have to be done in this PR.

Reply:
Hmm... maybe the reason for the static method was that the original author didn't like making a constructor throw an exception?